Fix permadiff that reorders  `stateful_external_ip` blocks on `google_compute_instance_group_manager` and `google_compute_region_instance_group_manager` resources

SarahFrench opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

Relates to (but doesn't close!) this issue : https://github.com/hashicorp/terraform-provider-google/issues/13430

This PR updates code related to `stateful_external_ip` blocks in the `google_compute_instance_group_manager` & `google_compute_region_instance_group_manager` resources to address a permadiff that proposes re-ordering those blocks. The root of this problem is that the API returns those blocks in a sorted order that doesn't match the user's config. The state will contain data returned from the API, so there's always a plan trying to return to the config-defined order.

Prior to this PR, the flattener function for `stateful_external_ip` blocks performs a conversion from a map (how the data from the API is presented by the compute Go client library) to an array (how the data is stored in state). 

This PR updates the flattener function to also re-order data returned by the API so it matches the user's Terraform configuration. This is achieved by passing the `schema.ResourceData` data for the resource into the flattener, to act as a source of information about the order of those fields in the user's config.


<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
compute: fixed a permadiff that reordered `stateful_external_ip` and `stateful_internal_ip` blocks on `google_compute_instance_group_manager` and `google_compute_region_instance_group_manager` resources
```

SarahFrench commented:
@rileykarson - I requested you as reviewer as you provided initial [feedback on the issue this PR relates to](https://github.com/hashicorp/terraform-provider-google/issues/13430#issuecomment-1412645376) about how the solution would be to re-order information from the API to match the user's config.

I feel a bit weird about this logic being in the flattener, but on the other hand the flattener was already performing a data transformation. If you think this reordering should be performed in a function separate to the flattener (or should be pulled into a separate function that's used in there, etc) please let me know!

Having said that, we're lucky that whatever we land on would be removed when the 'true' fix for this issue, changing the field to a Set, is added in a future major release.

SarahFrench commented:
Odd error in the generate diffs build:

> error: RPC failed; curl 16 Error in the HTTP2 framing layer

I'll try again

/gcbrun

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 369 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 369 insertions(+), 6 deletions(-))

SarahFrench commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_group_manager.go.erb:
```diff
@@ -1322,22 +1322,58 @@ func flattenStatefulPolicyStatefulInternalIps(statefulPolicy *compute.StatefulPo
+
+	orderedResult := make([]map[string]interface{}, len(configIpOrder))
+	unexpectedIps := []map[string]interface{}{}
+	for _, el := range initialResult {
```
Looking at this with fresh eyes, I don't think I need to have `initialResult` (that contains the data transformed map => slice) and also have the separate process to create `orderedResult` (that's produced by re-ordering `initialResult`). Instead I could re-order the data and convert it from map => slice at the same time.

Which would be easier to understand?

SarahFrench replied:
Refactored in https://github.com/GoogleCloudPlatform/magic-modules/pull/9577/commits/65785828b58fc8f57e788995a0ac211efa5418ab but I'm open to reverting that commit

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 370 insertions(+), 6 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 370 insertions(+), 6 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `3259` 
Passed tests `2926` 
Skipped tests: `332` 
Affected tests: `1` 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccRegionInstanceGroupManager_APISideListRecordering </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccRegionInstanceGroupManager_APISideListRecordering`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/770028e6-5b6f-4e48-89bf-5c2ae650abad/recording/TestAccRegionInstanceGroupManager_APISideListRecordering.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/770028e6-5b6f-4e48-89bf-5c2ae650abad/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/770028e6-5b6f-4e48-89bf-5c2ae650abad/recording) for each test

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report 
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 388 insertions(+), 7 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 388 insertions(+), 7 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `3262` 
Passed tests `2930` 
Skipped tests: `332` 
Affected tests: `0` 

$\textcolor{green}{\textsf{All tests passed in REPLAYING mode.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/babe0278-f3bb-4bce-afa8-ce9213b4e6d4/build-log/replaying_test.log)

SarahFrench commented:
@rileykarson could you please take a look at this when you have a chance? Thanks!

rileykarson commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_group_manager.go.erb:
```diff
@@ -1322,20 +1322,46 @@ func flattenStatefulPolicyStatefulInternalIps(statefulPolicy *compute.StatefulPo
 	}
-	return result
+	if len(unexpectedIps) > 0 {
+		// Additional IPs returned from API but not in the config are appended to the end of the slice
```
What would happen if I had [nic0, nic1] locally and got back (nic1, nic2) from the API? I don't think I saw a case covering this, but could have missed it.

SarahFrench replied:
There wasn't a case for this, and it uncovered a bug! I just rebased the branch but the new changes are in 
db91a96

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 429 insertions(+), 7 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 429 insertions(+), 7 deletions(-))

rileykarson commented on mmv1/third_party/terraform/services/compute/resource_compute_region_instance_group_manager_test.go.erb:
```diff
@@ -367,6 +367,28 @@ func TestAccRegionInstanceGroupManager_stateful(t *testing.T) {
+		CheckDestroy:             testAccCheckRegionInstanceGroupManagerDestroyProducer(t),
+		Steps: []resource.TestStep{
+			// {
+			// 	Config: testAccRegionInstanceGroupManager_statefulOrdered(context),
```
Should these lines + config block be deleted, or uncommented? I _think_ just testAccRegionInstanceGroupManager_statefulUnordered covers the interesting parts of the change.

SarahFrench replied:
I've deleted them, thanks for spotting it!

rileykarson commented on mmv1/third_party/terraform/services/compute/resource_compute_instance_group_manager_internal_test.go.erb:
```diff
@@ -0,0 +1,298 @@
+	}
+}
+
+func TestFlattenStatefulPolicyStatefulExternalIps(t *testing.T) {
```
note: just flagging I am only reviewing from this point in the file onwards, let me know if I'm missing anything.

SarahFrench replied:
Thanks - you're right to do that - the other tests in the file aren't changed but the diff is large just because of the file extension change.

rileykarson reviewed (approved):
Small note on deadcode in the handwritten test, but otherwise LGTM. Going through some snippets in the parent comments:

> The root of this problem is that the API returns those blocks in a sorted order that doesn't match the user's config.

Just to make sure I'm not missing something, it returns them in an _order_ that doesn't match the user's config, but not a _sorted_ order, i.e. the ordering on the API side does not matter, right? That's my understanding based on https://github.com/hashicorp/terraform-provider-google/issues/13430#issuecomment-1415504584

> I feel a bit weird about this logic being in the flattener...

Seems right to me. I'd only really consider breaking out of a flattener if we were accessing multiple same-level fields. We _could_ use a decoder to handle both the external and internal versions in the same place, although flatteners are fine since their individual handlers are still field-scoped. (wrt. child fields in when I'd use a flattener vs an alternative, a parent can access its children in its flattener)

> Also, should I extend this logic to also affect the [stateful_internal_ip](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_region_instance_group_manager#stateful_internal_ip) field?

I think so, unless there's a reason this issue couldn't happen there!

modular-magician commented:
#### Tests analytics 
Total tests: `860` 
Passed tests `795` 
Skipped tests: `65` 
Affected tests: `0` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>compute</li></ul></blockquote></details> 

$\textcolor{green}{\textsf{All tests passed in REPLAYING mode.}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/3910db89-1bb4-43b3-8fa4-f484d3a21d00/build-log/replaying_test.log)

SarahFrench commented:
Thanks for the review!

>>The root of this problem is that the API returns those blocks in a sorted order that doesn't match the user's config.
>
>Just to make sure I'm not missing something, it returns them in an order that doesn't match the user's config, but not a sorted order, i.e. the ordering on the API side does not matter, right? That's my understanding based on https://github.com/hashicorp/terraform-provider-google/issues/13430#issuecomment-1415504584

Yeah you're right that the API side ordering doesn't appear to matter. It's not clear if the reordering is due to the API explicitly choosing to do that or because the Compute Go client library storing the data in a map results in losing the old ordering. Below's an example of the map where the keys correspond to `interface_name` values in the TF configuration and the keys appear alphabetically ordered, which means slices generated from it are alphabetical and may not match a user's TF configuration:
 
 ```
 map[string]google.golang.org/api/compute/v1.StatefulPolicyPreservedStateNetworkIp {
    "nic0": {AutoDelete: "NEVER", ForceSendFields: []string len: 0, cap: 0, nil, NullFields: []string len: 0, cap: 0, nil},
    "nic1": {AutoDelete: "NEVER", ForceSendFields: []string len: 0, cap: 0, nil, NullFields: []string len: 0, cap: 0, nil},
}
 ```

 ---

> >Also, should I extend this logic to also affect the [stateful_internal_ip](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_region_instance_group_manager#stateful_internal_ip) field?
>
>I think so, unless there's a reason this issue couldn't happen there!

I'll update this PR to pull the sorting code into a reusable function and I'll update the `flattenStatefulPolicyStatefulInternalIps` func to also use it.

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report
Your PR generated some diffs in downstreams - here they are.

Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 373 insertions(+), 21 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-9577-old..auto-pr-9577) ( 4 files changed, 373 insertions(+), 21 deletions(-))

modular-magician commented:
#### Tests analytics 
Total tests: `860` 
Passed tests `794` 
Skipped tests: `65` 
Affected tests: `1` 

<details><summary>Click here to see the affected service packages</summary><blockquote><ul><li>compute</li></ul></blockquote></details> 

#### Action taken 
<details> <summary>Found 1 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests</summary><blockquote>TestAccRegionInstanceGroupManager_APISideListRecordering </blockquote></details> 

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$ 
 `TestAccRegionInstanceGroupManager_APISideListRecordering`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/084ee6bc-ff6e-42cc-b333-6b860b21b32f/recording/TestAccRegionInstanceGroupManager_APISideListRecordering.log)] 

##### Rerun these tests in REPLAYING mode to catch issues 

$\textcolor{green}{\textsf{No issues found for passed tests after REPLAYING rerun.}}$ 

---
$\textcolor{green}{\textsf{All tests passed!}}$ 
View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/084ee6bc-ff6e-42cc-b333-6b860b21b32f/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-9577/artifacts/084ee6bc-ff6e-42cc-b333-6b860b21b32f/recording) for each test

SarahFrench commented:
Re-requesting review as I've made big-ish changes since the last approval and don't want to abuse the fact the approval isn't dismissed by new commits
